### PR TITLE
[eclipse/xtext#1679] ported xtend code 2 java

### PR DESCRIPTION
**Pull request by cdietrich**
[eclipse/xtext#1679] ported xtend code 2 java
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.builder.standalone.tests/src/org/eclipse/xtext/builder/tests/MyGenerator.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.eclipse.xtext.generator.IFileSystemAccess;
+import org.eclipse.xtext.generator.IFileSystemAccess2;
+import org.eclipse.xtext.generator.IGenerator;
+import org.eclipse.xtext.xbase.lib.IteratorExtensions;
+
+import com.google.common.collect.Iterables;
+
+public class MyGenerator implements IGenerator {
+	@Override
+	public void doGenerate(final Resource input, final IFileSystemAccess fsa) {
```
Comment (tivervac) [inline org.eclipse.xtext.builder.standalone.tests/src/org/eclipse/xtext/builder/tests/MyGenerator.java]: Unneeded finals

File: org.eclipse.xtext.builder.standalone.tests/src/org/eclipse/xtext/builder/tests/MyGenerator.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		for (final Element ele : Iterables.filter(IteratorExtensions.toIterable(input.getAllContents()),
```
Comment (tivervac) [inline org.eclipse.xtext.builder.standalone.tests/src/org/eclipse/xtext/builder/tests/MyGenerator.java]: Don't we prefer `Streams.stream(...).filter(...)`?

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+									return new ByteArrayInputStream(bytes);
+								}
+							};
+						}
+					});
+				}
+			}
+			return super.findResource(path);
+		} catch (MalformedURLException e) {
+			throw Exceptions.sneakyThrow(e);
```
Comment (tivervac) [inline org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: We don't like sneakies

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+								}
+							};
+						}
+					});
+				}
+				return super.getResource(path);
+			}
+			return null;
+		} catch (MalformedURLException e) {
+			throw Exceptions.sneakyThrow(e);
```
Comment (tivervac) [inline org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: Sneaky

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	@Override
+	protected URL findResource(String path) {
+		try {
+			if (path.endsWith(".class")) {
+				String className = pathToClassName(path);
+				byte[] bytes = classMap.get(className);
+				if (bytes != null) {
+					return new URL("in-memory", null, -1, path, new URLStreamHandler() {
```
Comment (tivervac) [inline org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: You can extract this code, it's being used in the method above as well

File: org.eclipse.xtext.builder.standalone.tests/src/org/eclipse/xtext/builder/tests/MyGenerator.java (hunk 2 above)
Comment (cdietrich) [reply org.eclipse.xtext.builder.standalone.tests/src/org/eclipse/xtext/builder/tests/MyGenerator.java]: have no idea what we prefer

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 4 above)
Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: is lefft there by intention as it was there before

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: i want to keep it here as i dont want to break. see above

Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: what is your alternative in this case.
swallow the exception and do nothing?

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 4 above)
Comment (tivervac) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: It was there in the Java code you mean? I don't see it in the xtend code

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 3 above)
Comment (tivervac) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: Either return null (and log) or rethrow the error.

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 4 above)
Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: Well this code is as tend generated it before

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: Retro’s is what sneaky does. The api of classloaddr is bs. I don’t want to do some fancy different exception handler right now

Comment (tivervac) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: Okay

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java (hunk 4 above)
Comment (tivervac) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java]: Okay

**Pull request merged**